### ISSUE TITLE
tf.contrib.ffmpeg.decode_audio causes kernel crash w/ multi-threading

### ISSUE THREAD

**Issue by VinceMarron** [tensorflow/tensorflow#10196]
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow version (use command below)**: v1.1.0 - gpu
- **CUDA/cuDNN version**: 8.0
- **Exact command to reproduce**:

I have noticed that attempting to run the 'tf.contrib.ffmpeg.decode_audio' function with multiple threads causes the kernel to crash. This occurs when trying to create batches of data from audio binaries. 

The code underlying 'tf.contrib.ffmpeg.decode_audio' appears to a basic reference to functions outside of tensorflow so I am unsure there is a solution inside of the tensorflow domain. Nonetheless I wanted to bring this up in case someone had a solution. This is probably not a bug report and more of a low priority feature request. 

The code below will run without error when **num_threads=1** for tf.train.batch but the kernel will crash for **num_threads=2** or more. 

`
```
graph = tf.Graph()

with graph.as_default():

    
    batch_size=2

    queue = tf.train.slice_input_producer([paths, labels], 
                                          num_epochs=2, shuffle=True, capacity=32)
    
    
    audio_binary = tf.read_file(queue[0])
    signal = tf.contrib.ffmpeg.decode_audio(audio_binary, file_format='mp3', 
                                        samples_per_second=22500,  
                                        channel_count=1)[:450000,0]
    y_ = tf.one_hot(queue[1], 16, dtype=tf.float32)


    batch_sig, batch_y_ = tf.train.batch([signal, y_], batch_size=batch_size, 
                                         shapes=[(450000,), (16,)], 
                                         num_threads=1, capacity=64)
    
    
    with tf.Session(config=tf.ConfigProto(operation_timeout_in_ms=500)) as sess:
        sess.run(tf.global_variables_initializer())
        sess.run(tf.local_variables_initializer())

        
        coord = tf.train.Coordinator()
        threads = tf.train.start_queue_runners(sess=sess, coord=coord)
        
        
        for i in range(2):
            print(tf.reduce_max(batch_sig, axis=1).eval())
            
        coord.request_stop()
        coord.join(threads)            
        `

**Comment by douglaseck** [comment]
@fredbertsch can you have a look please?

**Comment by VinceMarron** [comment]
As an aside, I eventually found that `tf.train.slice_input_producer` is unlike the other input producers and does not return a queue object so the above code may be misleading. I am still confused as to the best way to handle queue/process/batch for lists of path/label pairs but the below seems to be closer:

`with graph.as_default():`
    
    batch_size=32
    n_samples = 100
    
    pathlist = tf.constant(paths, dtype=tf.string)[:n_samples]
    labellist = tf.constant(labels, dtype=tf.int32)[:n_samples]

    rsq = tf.RandomShuffleQueue(100000000, 0, [tf.string, tf.int32], shapes=[[],[]])
    do_enqueues = rsq.enqueue_many([pathlist, labellist])
        
    gotpath, gotlabel = rsq.dequeue()

    y_ = tf.one_hot(gotlabel, 16, dtype=tf.float32)

    signal = tf.contrib.ffmpeg.decode_audio(tf.read_file(gotpath), file_format='mp3', 
                                       samples_per_second=sr,  channel_count=1)[:450000,0]     

    batch_sig, batch_y_ = tf.train.batch([signal, y_], batch_size=batch_size, 
                                         shapes=[(450000,), (16,)], 
                                         num_threads=1, capacity=32)
    
    
    with tf.Session(config=tf.ConfigProto(operation_timeout_in_ms=5000)) as sess:

        sess.run(tf.global_variables_initializer())
        sess.run(tf.local_variables_initializer())

        
        coord = tf.train.Coordinator()
        threads = tf.train.start_queue_runners(sess=sess, coord=coord)
        sess.run(do_enqueues)
        
        i=0
        
        try:
            while not coord.should_stop():
                for step in range(n_samples//batch_size):
                    print(i, tf.reduce_max(batch_sig, axis=1).eval())
                    i=i+1
                coord.request_stop()
        except tf.errors.OutOfRangeError:
            coord.request_stop()
            print('Done training -- epoch limit reached')
        finally:
            coord.request_stop()
            coord.join(threads)            
`

**Comment by woodshop** [comment]
This issue appears to be the same as #5804, which has been open for a while now. Perhaps it's low-priority for the developers. 

However, I suggest that `ffmpeg.decode_audio` is not always a great solution for reading audio files, especially in large numbers. As you point out, decode_audio relies on the external ffmpeg library. In addition, for each audio file, it induces an additional disk write and read because it creates a temporary file during decoding. (According to #5804, the naming of these files is part of the multithreading problem.) For many thousands of audio files, that's a lot of disk IO. 

Instead, perhaps preprocess your audio files offline by decoding them (with ffmpeg or any other tool) and saving them into TFRecords. Decoded audio files can be encapsulated in `tf.train.SequenceExample` instances. Use the `SerializeToString` method to save a `SequenceExample` to a `TFRecord` using a `tf.python_io.TFRecordWriter` instance. Use a `tf.TFRecordReader` instance to read examples from the record and `tf.parse_single_sequence_example` to parse the examples.

**Comment by VinceMarron** [comment]
You are correct, it is the same issue as #5804. Closing here. 

Thank you for the advice.
